src/scenechange/mod.rs

```diff
@@ -104,237 +136,289 @@ impl SceneChangeDetector {
+    // ititiallization is different depending on frame set length
+    if self.deque_offset > 0
+      && frame_set.len() > self.deque_offset + 1
+      && self.score_deque.is_empty()
+    {
+      self.initialize_score_deque(
+        frame_set,
+        input_frameno,
+        previous_keyframe,
+        self.deque_offset,
+      );
+    } else if self.score_deque.is_empty() {
+      self.initialize_score_deque(
+        frame_set,
+        input_frameno,
+        previous_keyframe,
+        frame_set.len() - 1,
+      );
 
-  /// Determines if `current_frame` should be a keyframe.
-  fn is_key_frame<T: Pixel>(
-    &self, previous_frame: Arc<Frame<T>>, current_frame: Arc<Frame<T>>,
-    current_frameno: u64, previous_keyframe: u64,
-  ) -> bool {
-    if self.excluded_frames.contains(&current_frameno) {
-      return false;
+      self.deque_offset = frame_set.len() - 2;
+    }
+    // Running single frame comparison and adding it to deque
+    // Decrease deque offset if there is no new frames
+    if frame_set.len() > self.deque_offset + 1 {
+      self.run_comparison(
+        frame_set[self.deque_offset].clone(),
+        frame_set[self.deque_offset + 1].clone(),
+        input_frameno,
+        previous_keyframe,
+      );
+    } else {
+      self.deque_offset -= 1;
     }
 
-    let result = self.has_scenecut(
-      previous_frame,
-      current_frame,
-      current_frameno,
-      previous_keyframe,
-    );
+    // Adaptive scenecut check
+    let scenecut = self.adaptive_scenecut();
     debug!(
-      "[SC-Detect] Frame {} to {}: I={:.3} T={:.3} P={:.3} {}",
-      current_frameno - 1,
-      current_frameno,
-      result.intra_cost,
-      result.threshold,
-      result.inter_cost,
-      if result.has_scenecut { "Scenecut" } else { "No cut" }
+      "[SC-Detect] Frame {}: I={:4.0}  T= {:.0} {}",
+      input_frameno,
+      self.score_deque[self.deque_offset].0,
+      self.score_deque[self.deque_offset].1,
+      if scenecut { "Scenecut" } else { "No cut" }
     );
-    result.has_scenecut
-  }
 
-  /// Uses lookahead to avoid coding short flashes as scenecuts.
-  /// Saves excluded frame numbers in `self.excluded_frames`.
-  fn exclude_scene_flashes<T: Pixel>(
-    &mut self, frame_subset: &[Arc<Frame<T>>], frameno: u64,
-    previous_keyframe: u64,
-  ) {
-    let lookahead_distance = self.lookahead_distance;
-
-    if frame_subset.len() - 1 < lookahead_distance {
-      // Don't add a keyframe in the last frame pyramid.
-      // It's effectively the same as a scene flash,
-      // and really wasteful for compression.
-      for frame in frameno..=(frameno + lookahead_distance as u64) {
-        self.excluded_frames.insert(frame);
+    if scenecut {
+      // Clear buffers and deque
+      self.frame_buffer.clear();
+      debug!("[SC-score-deque]{:.0?}", self.score_deque);
+      self.score_deque.clear();
+    } else {
+      // Keep score deque 5 + lookahead_size frames
```

Comment:
I wonder how did we reach 5+ lookahead_size, would be nice to add in commit message or here, whichever is better.

On seeing further code, could see the mention of 5 logic, would be better if it is mentioned here than later.

Reply:
IIRC 5 was chosen because it's the size of a frame pyramid. I think at one point there was at least one comment mentioning that, although I haven't touched this code in a while so my memory is fuzzy.

Reply:
Improved comment